server/s3_list_object_versions_action.cc

```diff
@@ -0,0 +1,835 @@
+    if (max_record_count <= 1) {
+      set_s3_error("InternalError");
+      fetch_successful = false;
+      s3_log(S3_LOG_ERROR, request_id,
+             "Next keyval operation failed due rpc message size threshold, "
+             "even with max record retrieval as one\n");
+    } else {
+      retry_count++;
+      s3_log(S3_LOG_INFO, request_id, "Retry next key val, retry count = %d\n",
+             retry_count);
+      get_next_versions();
+      s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+      return;
+    }
+  } else {
+    s3_log(S3_LOG_DEBUG, request_id, "Failed to find versions listing\n");
+    set_s3_error("InternalError");
+    fetch_successful = false;
+  }
+  send_response_to_s3_client();
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3ListObjectVersionsAction::check_latest_versions() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  std::shared_ptr<S3ObjectMetadata> version_metadata = nullptr;
+  size_t version_list_offset_temp = 0;
+  // Run through version_list, check and set "latest" flag.
+  for (auto&& version : versions_list) {
+    if (last_object_checked.empty()) {
+      version_metadata = version;
+      break;
+    } else if ((version_list_offset_temp <= version_list_offset) ||
+               (last_object_checked == version->get_object_name())) {
+      version_list_offset_temp++;
+      continue;
+    } else {
+      version_metadata = version;
+      break;
+    }
+  }
+  if (version_metadata != nullptr) {
+    version_list_offset = version_list_offset_temp;
+    last_object_checked = version_metadata->get_object_name();
+    const auto& object_list_idx_lo =
+        bucket_metadata->get_object_list_index_layout();
+    const auto& obj_version_list_idx_lo =
+        bucket_metadata->get_objects_version_list_index_layout();
+
+    if (zero(object_list_idx_lo.oid) || zero(obj_version_list_idx_lo.oid)) {
+      // Object list index and version list index missing.
+      get_next_versions_failed();
+    } else {
+      // Read object metadata from object index.
+      object_metadata = object_metadata_factory->create_object_metadata_obj(
+          request, bucket_name, version_metadata->get_object_name(),
+          object_list_idx_lo, obj_version_list_idx_lo);
+      object_metadata->load(
+          std::bind(&S3ListObjectVersionsAction::fetch_object_info_success,
+                    this),
+          std::bind(&S3ListObjectVersionsAction::get_next_versions_failed,
+                    this));
+    }
+  } else {
+    next();
+  }
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3ListObjectVersionsAction::fetch_object_info_success() {
+  // Compare the versions and set the latest flag
+  std::shared_ptr<S3ObjectMetadata> version_metadata =
+      versions_list[version_list_offset];
+  if (object_metadata->get_obj_version_id() ==
+      version_metadata->get_obj_version_id()) {
+    version_metadata->set_latest(true);
+  }
+  check_latest_versions();
```

Comment:
Not sure I understand how this works. Why is this called here, but it's also designated as an Action Step?

Reply:
A version_list that we fetched could have multiple versions of multiple objects. The check_latest_versions() function will be called for checking if the first version of the object in the version_list is the latest or not.
So, we start checking the above after we are done fetching the version_list. And we need to keep calling check_latest_versions() till all the objects within the fetched version_list are checked for the latest.